scripts/artifacts.py

```diff
@@ -0,0 +1,263 @@
+
+    plt.tight_layout()
+    plt.savefig(out)
+
+
+@cli.command()
+@click.option("--previous", required=True)
+@click.option("--next", required=True)
+def compare_stats(previous, next):
+    previous_stats = mk_commit_stats(fetch_commit_info(previous))
+    next_stats = mk_commit_stats(fetch_commit_info(next))
+
+    print_diffs(
+        mk_diffs(previous_stats.gas_costs, next_stats.gas_costs),
+        title="Gas costs",
+        header_previous=previous_stats.info.ref,
+        header_next=next_stats.info.ref,
+    )
+
+    print_diffs(
+        mk_diffs(previous_stats.entrypoint_sizes, next_stats.entrypoint_sizes),
+        title="Entrypoint sizes",
+        header_previous=previous_stats.info.ref,
+        header_next=next_stats.info.ref,
+    )
+
+
+# From the given $ref (usually a commit hash or a branch name), fetch the latest $limit commits
+# (from children to parents) and their workflow ids.
+def fetch_commit_infos(ref, *, limit):
+    query = """
+    query($ref:String!, $limit:Int!) {
+      repository(owner: "tezos-checker", name: "checker") {
+        object(expression: $ref) {
+          ... on Commit {
+            history(first: $limit) {
+              nodes {
+                message
+                oid
+                statusCheckRollup {
+                  contexts(first: 10) {
+                    edges {
+                      node {
+                        ... on CheckRun {
+                          checkSuite {
+                            workflowRun {
+                              databaseId
+                            }
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  """
+
+    vars = {"ref": ref, "limit": limit}
+
+    ret = requests.post(
+        ENDPOINT_GRAPHQL, json={"query": query, "variables": vars}, headers=headers
+    )
+    ret.raise_for_status()
+    ret = ret.json()
+
+    commit_infos = [
+        CommitInfo(
+            message=get_first_line(commit["message"]),
+            ref=ref if i == 0 else f"{ref}~{i}",
+            rev=commit["oid"],
+            workflow_ids=[
+                check["node"]["checkSuite"]["workflowRun"]["databaseId"]
+                for check in commit["statusCheckRollup"]["contexts"]["edges"]
+            ]
+            if commit["statusCheckRollup"]
+            else [],
+        )
+        for i, commit in enumerate(
+            ret["data"]["repository"]["object"]["history"]["nodes"]
+        )
+        if commit["statusCheckRollup"]
+    ]
+
+    return commit_infos
+
+
+def fetch_commit_info(expr):
+    return fetch_commit_infos(expr, limit=1)[0]
+
+
+def mk_commit_stats(info):
+    stats = download_stats(info.workflow_ids)
+
+    gas_costs = stats.get("gas-costs", [{}])[0]
+    # TODO: Gas costs are strings for no reason.
+    gas_costs = {k: int(v) for k, v in gas_costs.items()}
```

Comment:
True :thinking: Hmmm, from [what I can see](https://pytezos.org/quick_start.html?highlight=gas_limit#reveal-public-key) representing all items as strings is a PyTezos choice (including `gas_limit`). We could change our own generator to convert it to an int before we store the information in the file ([here](https://github.com/tezos-checker/checker/blob/4d1e384af141f86ce7957342a657320d2b430935/e2e/main.py#L266)) but then the past gas-cost files would become ill-formed. I'd probably leave it as is for now

Reply:
I think there's a way to do it gradually. If we change the E2E tests to emit `int`'s, then this cast will keep working. After a while, when `master` has the new version for a while, we can remove the now redundant `int` cast.

Reply:
Ah, right, great thinking! :ok_hand: